src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.cs

```diff
@@ -0,0 +1,306 @@
+            {
+                context.RegisterSyntaxNodeAction(c => AnalyzeIfEnabled(c, analyzeSyntax), syntaxKinds);
+            }
+
+            // check for all the other things tht can be registered as well.
+        }
+
+        private void AnalyzeIfEnabled(CodeBlockAnalysisContext context, Action<CodeBlockAnalysisContext> analyze)
+            => AnalyzeIfEnabled(context, analyze, context.Options, context.SemanticModel.SyntaxTree, context.CancellationToken);
+
+        private void AnalyzeIfEnabled(SyntaxNodeAnalysisContext context, Action<SyntaxNodeAnalysisContext> analyze)
+            => AnalyzeIfEnabled(context, analyze, context.Options, context.SemanticModel.SyntaxTree, context.CancellationToken);
+
+        private void AnalyzeIfEnabled<TContext>(
+            TContext context, Action<TContext> analyze, 
+            AnalyzerOptions options, SyntaxTree syntaxTree, CancellationToken cancellationToken)
+        {
+            if (!AnalyzerShouldRun(options, syntaxTree, cancellationToken))
+            {
+                return;
+            }
+
+            analyze(context);
+        }
+
+        private bool AnalyzerShouldRun(AnalyzerOptions options, SyntaxTree syntaxTree, CancellationToken cancellationToken)
+        {
+            var optionSet = options.GetDocumentOptionSetAsync(syntaxTree, cancellationToken).GetAwaiter().GetResult();
+            if (optionSet == null)
+            {
+                return false;
+            }
+
+            var optionValue = optionSet.GetOption(_option);
+            switch (optionValue.Notification.Severity)
+            {
+                case ReportDiagnostic.Error:
+                case ReportDiagnostic.Warn:
+                case ReportDiagnostic.Info:
+                    return true;
+            }
+
+            return false;
+        }
+
+        public abstract class CodeStyle2DiagnosticAnalyzer : AbstractCodeStyleDiagnosticAnalyzer
+        {
+            public readonly TCodeStyleProvider _codeStyleProvider;
+
+            protected CodeStyle2DiagnosticAnalyzer(bool configurable = true) 
+                : this(new TCodeStyleProvider(), configurable)
+            {
+            }
+
+            private CodeStyle2DiagnosticAnalyzer(TCodeStyleProvider codeStyleProvider, bool configurable)
+                : base(codeStyleProvider._descriptorId, codeStyleProvider._title, codeStyleProvider._message, configurable)
+            {
+                _codeStyleProvider = codeStyleProvider;
+            }
+
+            protected sealed override void InitializeWorker(AnalysisContext context)
+                => _codeStyleProvider.DiagnosticAnalyzerInitialize(context);
+
+            public sealed override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+                => _codeStyleProvider.GetDiagnosticAnalyzerCategory();
+
+            public sealed override bool OpenFileOnly(Workspace workspace)
+                => _codeStyleProvider.DiagnosticsForOpenFileOnly(workspace);
```

Comment:
all members in these classes are intentionally sealed.  All logic should be provided in the CodeStyleProvider subclass.